File: java/1721.0_Swapping_Nodes_in_a_Linked_List.java
/*
Runtime: 3 ms, faster than 81.58% of Java online submissions for Swapping Nodes in a Linked List.
Memory Usage: 182.8 MB, less than 31.99% of Java online submissions for Swapping Nodes in a Linked List.
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode n1 = null, n2 = null;
        for (var p=head; p != null; p = p.next) {
            if (n2 != null) {
                n2 = n2.next;
            }
            if (--k == 0) {
                n1 = p;
                n2 = head;
            }
        }
        // swap
        int tmp = n1.val;
        n1.val = n2.val;
        n2.val = tmp;
        return head;
    }
}


/*
Runtime: 4 ms, faster than 51.04% of Java online submissions for Swapping Nodes in a Linked List.
Memory Usage: 188.5 MB, less than 10.61% of Java online submissions for Swapping Nodes in a Linked List.
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode n1 = null, n2 = null;
        for (var p=head; p != null; p = p.next) {
            n2 = n2 == null ? n2: n2.next;
            if (--k == 0) {
                n1 = p;
                n2 = head;
            }
        }
        // swap
        var tmp = n1.val;
        n1.val = n2.val;
        n2.val = tmp;
        return head;
    }
}


/*
Runtime: 2 ms, faster than 100.00% of Java online submissions for Swapping Nodes in a Linked List.
Memory Usage: 178.8 MB, less than 56.73% of Java online submissions for Swapping Nodes in a Linked List.
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode n1 = null, n2 = null;
        for (ListNode p=head; p != null; p = p.next) {
            n2 = n2 == null ? n2: n2.next;
            if (--k == 0) {
                n1 = p;
                n2 = head;
            }
        }
        // swap
        int tmp = n1.val;
        n1.val = n2.val;
        n2.val = tmp;
        return head;
    }
}

File: cpp/1721.0_Swapping_Nodes_in_a_Linked_List.cpp
/*
one pass
T: O(N)
S: O(1)

Runtime: 749 ms, faster than 77.85% of C++ online submissions for Swapping Nodes in a Linked List.
Memory Usage: 180.2 MB, less than 57.49% of C++ online submissions for Swapping Nodes in a Linked List.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        ListNode *n1 = nullptr, *n2 = nullptr;
        for (auto p = head; p != nullptr; p = p->next) {
            if (n2 != nullptr) {
                n2 = n2->next;
            }
            if (--k == 0) {
                n1 = p;
                n2 = head;
            }
        }
        swap(n1->val, n2->val);
        return head;
    }
};


/*
Runtime: 879 ms, faster than 51.28% of C++ online submissions for Swapping Nodes in a Linked List.
Memory Usage: 180.2 MB, less than 15.97% of C++ online submissions for Swapping Nodes in a Linked List.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        ListNode *n1 = nullptr, *n2 = nullptr;
        for (auto p = head; p != nullptr; p = p->next) {
            n2 = n2 == nullptr ? n2: n2->next;
            if (--k == 0) {
                n1 = p;
                n2 = head;
            }
        }
        swap(n1->val, n2->val);
        return head;
    }
};

